src/dfx/src/commands/new.rs

```diff
@@ -196,7 +197,32 @@ fn write_files_from_entries<R: Sized + Read>(
         });
 
         let p = PathBuf::from(p);
-        create_file(log, p.as_path(), &v, dry_run)?;
+        if p.extension() == Some("json-patch".as_ref()) {
+            let patch: json_patch::Patch = serde_json::from_slice(&v)
+                .with_context(|| format!("Failed to parse {}", p.display()))?;
+            let to_patch = p.with_extension("json");
+            ensure!(
+                to_patch.exists(),
+                "Failed to patch {}: not found",
+                to_patch.display()
+            );
+            let mut value = load_json_file(&to_patch)?;
+            json_patch::patch(&mut value, &patch)
+                .with_context(|| format!("Failed to patch {}", to_patch.display()))?;
+            save_json_file(&to_patch, &value)?;
+        } else if p.extension() == Some("patch".as_ref()) {
+            let v = std::str::from_utf8(&v)
+                .with_context(|| format!("Failed to parse {}", p.display()))?;
+            let patch = patch::Patch::from_single(v)
+                .map_err(|e| anyhow!("Failed to parse {}: {e}", p.display()))?;
+            let to_patch = p.with_extension("");
+            let existing_content = dfx_core::fs::read_to_string(&to_patch)?;
+            let patched_content = apply_patch::apply_to(&patch, &existing_content)
+                .with_context(|| format!("Failed to patch {}", to_patch.display()))?;
+            dfx_core::fs::write(&to_patch, &patched_content)?;
```

Comment:
suggestion: extract method (and this will make it more obvious, as @sesi200 noted, that `dry_run` isn't taken into account)